neutron/services/loadbalancer/agent_scheduler.py

```diff
@@ -93,6 +96,73 @@ def update_lbaas_agent_hosting_pool(self, context, id, agent):
             query[0].agent = agent
             context.session.flush()
 
+    def _check_if_pool_has_added(self, context, pool_id):
+        query = context.session.query(PoolLoadbalancerAgentBinding)
+        query = query.filter(PoolLoadbalancerAgentBinding.pool_id == pool_id)
+        try:
+            binding = query.one()
+        except exc.NoResultFound:
+            binding = None
+        if binding:
+            raise loadbalancer.PoolHasBoundAgent(pool=pool_id,
+                                                 agent=binding.agent_id)
+
+    def _bind_pool(self, context, agent_id, pool_id):
+        with context.session.begin(subtransactions=True):
+            binding = PoolLoadbalancerAgentBinding()
+            binding.agent_id = agent_id
+            binding.pool_id = pool_id
+            context.session.add(binding)
+
+    def add_pool_to_lbaas_agent(self, context, agent_id, pool_id):
+        agent = self._get_agent(context, agent_id)
+        pool = self.get_pool(context, pool_id)
+        # check if pool could add to the agent
+        self._check_if_pool_has_added(context, pool_id)
+        self._bind_pool(context, agent_id, pool_id)
+        driver = self._get_driver_for_provider(pool['provider'])
+        driver.add_pool_to_agent(context, pool, agent)
+
+    def _unbind_pool(self, context, agent, pool):
+        with context.session.begin(subtransactions=True):
+            query = context.session.query(PoolLoadbalancerAgentBinding)
+            query = query.filter(
+                PoolLoadbalancerAgentBinding.pool_id == pool,
+                PoolLoadbalancerAgentBinding.agent_id == agent)
+            try:
+                binding = query.one()
+            except exc.NoResultFound:
+                raise loadbalancer.PoolNotBoundAgent(pool=pool, agent=agent)
+            context.session.delete(binding)
+
+    def _get_same_pools_by_pool_id(self, context, pool_id):
+        pool = self.get_pool(context, pool_id)
+        pools = [pool_id]
+        with context.session.begin(subtransactions=True):
+            try:
+                vip = self._get_resource(context, ldb.Vip, pool['vip_id'])
+            except loadbalancer.VipNotFound:
+                # in this case, pool not bound to a vip
+                return set(pools)
+            for _vip in vip.port.vips:
+                pools.append(_vip.pool_id)
+        return set(pools)
+
+    def remove_pool_from_lbaas_agent(self, context, agent_id, pool_id):
```

Comment:
那就在add的时候，判断这个pool对应的vip下的其它pool是否在同一个agent， 如果不是 就raise个异常 不让add

2016-07-12 14:50 GMT+08:00 huntxu <notifications@github.com>:

> In neutron/services/loadbalancer/agent_scheduler.py
> <https://github.com/eayunstack/neutron/pull/39#discussion_r70385167>:
>
> > +            context.session.delete(binding)
> > +
> > +    def _get_same_pools_by_pool_id(self, context, pool_id):
> > +        pool = self.get_pool(context, pool_id)
> > +        pools = [pool_id]
> > +        with context.session.begin(subtransactions=True):
> > +            try:
> > +                vip = self._get_resource(context, ldb.Vip, pool['vip_id'])
> > +            except loadbalancer.VipNotFound:
> > +                # in this case, pool not bound to a vip
> > +                return set(pools)
> > +            for _vip in vip.port.vips:
> > +                pools.append(_vip.pool_id)
> > +        return set(pools)
> > +
> > +    def remove_pool_from_lbaas_agent(self, context, agent_id, pool_id):
>
>
> 这也是为什么我上面提到vip的问题，如果用户操作失误将pool添加到其他agent，那就是添加应该失败，而不应该当删一个pool的时候将其他的pool也从agent中去掉。除非我们改变整个策略以vip来调度所有连接着同一个vip的pool才能够这么做，但这样只对haproxy这个driver有效，还得单独做处理，麻烦。
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eayunstack/neutron/pull/39/files/609dd87c1c005386d538a761bbb1b4b09842b8c4#r70385167>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGUT0YNzLHtJYexATVs-dpB5fXFTLWmcks5qUzkqgaJpZM4JJO13>
> .
>